### ISSUE TITLE
Consistent value for name attribute in datasets

### ISSUE THREAD

**Issue by Bultako** [gammapy/gammapy#2517]
We have a `name` attribute for 1D and 3D datasets. The values for these attributes are set in:

https://github.com/gammapy/gammapy/blob/fe6f23d8b7e3fc1d2f7f00a2dbe368769bea0d0d/gammapy/cube/make.py#L355

https://github.com/gammapy/gammapy/blob/fe6f23d8b7e3fc1d2f7f00a2dbe368769bea0d0d/gammapy/spectrum/make.py#L186

Is it intentional e.g. for fitting maps and spectra together? then maybe the naming should be a bit clearer...

**Comment by adonath** [comment]
> Is it intentional e.g. for fitting maps and spectra together?

No, I think this just happened by accident. Are there any preferences for the default name we should choose? I guess we could also just just enumerate the datasets and put e.g. `"map-dataset-1"` or `"spectrum-dataset-1"`. I think in general it's probably not a good idea to keep track of meta information such as observation id in the name, so maybe we shouldn't even try at all to do it in the name and introduce `.meta` later...

**Comment by adonath** [comment]
Note: a simple enumeration of the datasets (instead of the obs-id) also avoids this issue: https://github.com/gammapy/gammapy/issues/2515 on the dataset level.

**Comment by cdeil** [comment]
The question of names for models and datasets has come up many times in the past year. It's a difficult one, but it is solvable once and for all by making a few decisions and writing a few lines of documentation, so we should do that to avoid going in circles.

My 2 cents: generating good names for models and datasets is hard, because of different use cases. For the case of run-wise datasets, using the OBS_ID to make a name is the best solution. If you make a lightcurve and either split one run into many datasets, or combine different runs into one stacked dataset, then you're in trouble with the OBS_ID.

Possible solutions here are:
- Delete `name` attribute from dataset and model objects, because it's not needed at all, it's just "annotation" which is nice to have in some cases, but causing problems in others.
- If you keep the names, make a decision and state it in the docs whether Gammapy internally requires names to be unique. I'm strongly -1 to require unique names, because as far as i can see it's just not needed. (e.g. for compound models I removed the re-exposition of the sub-model parameters on the compound model, to avoid the issue of name clashes if e.g. multiple `norm` parameters are around).
- If unique names are required, it should be the Gammapy responsibility to generate those. The easiest solution would probably be to use a default `name = hex(id(self))` which is the Python way of generating a unique ID. I think it's what Tensorflow used to do to get unique names for tensors (very similar problem to Gammapy datasets), and now they have a global counter - that's easy to implement by putting an int at the base `Dataset` and `Model` class level, and then incrementing it in `__init__` for all models and datasets. I think this would make parallelisation harder in the future, but if the distributed solution has a "master" process where all datasets and models are created before the distributed part starts (as opposed to distributed start-up of worker processes), then it works.
- If no unique name is required, then probably putting the string "none" or "n/a" for (not available) or "default" is a better solution than OBS_ID, or my preference would still be to use `name = hex(id(self))`.  I've already put this into the `Parameter` repr (see [here](https://github.com/gammapy/gammapy/blob/d6ddf42cdd05893b46a3c232e9d2ef0cec1636f9/gammapy/modeling/parameter.py#L191)), because I found it necessary to debug the new model code, to see when a Parameter object is the same object, or a copy with the same values. It's not very pretty for users, but also not uncommon, most Python classes that don't define a repr show that object ID hex in the repr, Python users see it all the time.

- A solution that relies on users to always set names is annoying, and one that requires users to put unique names is terrible (even if we generate errors, the errors can only be generated later, once the models and datasets container objects are generated). Overall, we need to have a sentence or two like http://xarray.pydata.org/en/stable/faq.html#what-is-your-approach-to-metadata in the docs where we state to users how names work and who's responsible.

@adonath - Could you please take care of this and send a PR with your proposal?

Or if you prefer, I could also do it. To summarise: I'm OK to delete names, but I'm also OK to keep them. If we keep them, I would suggest to state that they don't have to be unique (and to add tests for that, like I already partially did for compound model a few days ago, but do it for skymodels and datasets), and to put `"none"` or `hex(id(self))` as default (in `__init__` in the base class).

**Comment by Bultako** [comment]
Do we need identifiers for datasets? If yes, they should be unique.
Names are not identifiers, so I'm ok with having duplicated names if we keep them, but do we really need names for datasets?

I think we should focus on the needs.
I guess the `name` attribute was introduced as a shortcut to access **individual** datasets. In that case I'd suggest to change the `name` attribute by an `id` attribute, with unique values generated by Gammapy, `hex(id(self))`is fine. Users should not modify the value of the `id`.

**Comment by cdeil** [comment]
@adonath @QRemy (and maybe @registerrier) - please comment what you propose we put concerning names / IDs for models and datasets in Gammapy.

If it's not clear what to put - I'd suggest a short call to discuss & decide ASAP.

**Comment by registerrier** [comment]
I suppose having a `name` attribute might be useful when dealing with multiple types of datasets, and then a speficic `id` not controlled by the user e.g. for a multi instrument analysis like joint crab, I can have my HESS datasets `hess-1`, `hess-2` etc then `fermi-1`  and `magic-1` etc. It might be a bit simpler for the user to identify which is which.

I can imagine that having to follow multiple `id` counters per existing `name` is not so nice.

**Comment by Bultako** [comment]
@registerrier 
This is storing provenance info in the datasets (which I'm not against), and for this I think we should do an exercise on what to store in a `meta` dictionary based on common use cases. In the case you expose it should be the telescope name, also a counter and the data reduction type (joint/stacked)? Once all this is done we should also add a helper method to access datasets based on the info stored.  But please do not make `name` values a mix of all provenance info. 

More questions:
Should we allow users to add/edit provenance info and provide them with an easy way to do it or it should be the responsibility of Gammapy? Do we have a short list of use cases or we do not have any control on that list? Is all this really needed or it can be easily done in another way with Python outside Gammapy?

**Comment by registerrier** [comment]
No this is not provenance, this is a name chosen by the user. One could consider that `obs_id` is an information more connected with provenance.

We shouldn't mix provenance information and analysis related information which can be needed by the user for analysis purpose.

**Comment by Bultako** [comment]
> No this is not provenance, this is a name chosen by the user.

Ok. 
This approach is much more flexible, filling user annotations it's not the responsibility of Gammapy but of the user. Note you can already do this with Python:
```
for i, ds in enumerate(datasets.datasets): 
    ds.telescope=f"HESS-{i}" 
```  
If we want to dig deeper, I see user annotations as non-fixed keyword structure of a `notes` dict attribute, then I guess we could provide a helper method to access/filter datasets from a dataset list based on the content of `notes`?

**Comment by adonath** [comment]
The way we use the `name` attribute for `SkyModel` and `Dataset` objects now is mostly to have convenience access in `SkyModels` and `Datasets` containers, so that you can write `models["my_model"]` and `datasets["my_dataset"]`. For models we also use it in the `FluxPointEstimator` to specify for which model component you want to compute the flux points for. For this use cases we rely on the names being unique and I think there is now way around this. The `name` attribute is clearly not the place to store provenance information, but it's a string alias for a model or dataset objects mostly to allow `dict` like access by key. Personally I think this is nicer than working with integer indices or passing python objects itself (but it's clearly a matter of taste). The whole point of the `name` attribute is to provide a human friendly string alias for a complex object.  

In general it can be optional to provide names to get the convenient access, but we don't have to rely on it in the API or generate default names. I think in some cases there are good reason to do it also by default. Imaging a `SkyModels` object with >10 components created from a catalog. It's just natural to use it like `models["crab-nebula"]` and get back the corresponding `SkyModel` object.

For datasets it's clear that we need more. Probably even a special data-structure, that keeps track of the event selection and telescope information. Each (IACT) dataset needs at least an `instrument_id`, `obs_id`, `event_type_id`, `time_interval_id` and `phase_interval_id`. I would propose even a specialised `DatasetID` data-structure for this or in the simplest case it can also be a meta dict. The combination of the IDs makes each datasets unique allows for grouping and selecting datasets, by observations, event types, phase and time intervals. Ideally you get a an overview table like so:

![Screenshot 2019-11-07 at 10 51 53](https://user-images.githubusercontent.com/3715928/68378564-9eb49b00-014c-11ea-8ae5-2955353df098.png)

Or similar for splitting e.g. one observation into multiple time intervals:

![Screenshot 2019-11-07 at 10 52 31](https://user-images.githubusercontent.com/3715928/68378623-b55af200-014c-11ea-8497-0977a970101a.png)

But this `DatasetID` or `DatasetsIndex` is completely independent of the question of the `name`
attribute.

**Comment by cdeil** [comment]
@adonath - Can you make a proposal how to resolve this issue, please?

> For datasets it's clear that we need more.

I'm not convinced.

I strongly suggest to not add these things now. Even with model and dataset names, I feel like if we hadn't added them, we would be in a better place than now, where we have to argue what to put for the names. Introducing many more IDs will be the same: suggest to only introduce it when the need is clear and the semantics are clear: do they have to be unique, and in what scope? is it user or Gammapy responsibility or mixed?

Adding provenance information is fine. The usual way is to add `meta` dicts and put information there what happened.

That's different from having IDs and using them in analysis. Usually you'll have all kind of data reductions that generate datasets. E.g. you can split by event type, or by time, or .... The beauty is that all you require is that the datasets are statistically independent, the same event only appears in one dataset, and that data reduction produces attached GTI and IRF that are appropriate to that selection. You don't need a "database" where you fetch EVENTS and IRF objects and match and join objects by IDs. Now e.g. LSST does use a database during data reduction and analysis, it's a way one can go, but I think for CTA it's just not needed, there we can do e.g. what Fermi-LAT does with event types, or HESS with runs: just generate multiple datasets, and require that the data reduction software spits out self-consistent datasets. You don't need a name (HESS or MAGIC), or an ID (event type XY), that's "only" provenance information, not analysis-required information.

The idea of managing the analysis-required informations was to create https://docs.gammapy.org/0.14/api/gammapy.data.ObservationFilter.html at the start, and then use it during data reduction and modeling & fitting, and if needed modify it together with the data as you go along through the data reduction chain. Maybe adding provenance infos like instrument name there is a good idea? Then `datasets.get(instrument_name="hess")` or other queries could defer to a query methods on `ObservationFilter` to satisfy common use cases to do selections, and we would get rid of the dataset name issue in a sense, although of course implementing and fleshing out that new more powerful (because not reduced to a single string) datasets selection scheme is work, but that could just be deferred to next year or be kept minimal for now.

**Comment by adonath** [comment]
@cdeil We're not that far apart. If you call it provenance information and you can still use it later to select e.g. a subset of datasets, this is also what I had in mind. The main purpose is that users can keep track of the datasets, during an interactive fitting and load and inspect (lots) of datasets after data reduction. Ideally with something like a `DatasetsIndex` or `DatasetStore`, which provides lazy loading as well. I don't necessarily agree on the statistical independence and limitation to choose everything at the beginning of the data reduction. Why not group and e.g. stack datasets on the fly?  You choose e.g. a fine time binning in the data reduction and then datasets are re-grouped by the `LightCurveEstimator`, given a larger time binning, or the background normalisation is linked for all datasets corresponding to one `obs-id`. Or users can decide to throw away one event type later in the analysis etc. This flexibility in the grouping and selection of the datasets later in the analysis seems very attractive to me and worth pursuing. But yes, this is of course work for next year.

> Even with model and dataset names, I feel like if we hadn't added them, we would be in a better place than now, where we have to argue what to put for the names.

I don't agree. I don't really see a good, user-friendly (interactive) handling of many model components, if you can't refer to them by name. If you use  e.g. `SkyModels.from_yaml("models.yaml")`, how can you ever see what it really contains, when there are no source identifiers? Guessing from the position or morphology? Both `astropy` and `sherpa` have model names as well. It's an established concept. 

My proposal here would be, that we don't rely on the `name` attributes in the API, but offer it optionally for convenience access of models and datasets, if it's defined by the user. From the Gammapy side we set it by default, wherever it makes sense (in `SourceCatalogObject.sky_model` or for dataset creation in our tutorials). In the `__init__` of `SkyModels` and `Datasets` we check for the uniqueness of the names, if they are provided and throw an error if there are duplications. 

The needed changes I see are:
- Add uniqueness check in `Datasets`
- Allow `Datasets.__getitem__` and `SkyModels.__getitem__` to take, `int`, `str` or `object`. If `object` is used the same object is just returned. I think this is what we do in `Parameters.__getitem__` already. The purpose of the `object` access is to just provide a uniform API, so that you can pass e.g. `FluxPointsEstimator(source=source_model)`.
- Find a workaround for `SpectrumDatasetOnOff.to_ogip_files()`, where I think the `OBS_ID` header keyword is generated from the `.name` attribute.
- Remove the default names in `SpectrumDatasetMaker` and `MapDatasetMaker` and just set the names outside in the observation loop using e.g. `dataset.name = f"{obs_id}"`. I don't really care what we choose there, a simple index such as `map-dataset-1` will do as well.

In summary: if users provide a name to the model or dataset, then they can access it by the name. They are free in the choice of the name (they can call there dataset `"ernie"` or `"bert"`), nowhere in the code we rely on the name being e.g. an `obs-id`. If there is no name, access happens by `int` or `object`.

**Comment by adonath** [comment]
One more change:
- Remove default name `"source"` in `SkyModel` and make the `source` argument in `FluxPointEstimator` and `LightCurveEstimator` mandatory.

**Comment by cdeil** [comment]
> Why not group and e.g. stack datasets on the fly?

That's fine, you can do that, you don't need names or IDs.

> This flexibility in the grouping and selection of the datasets later in the analysis seems very attractive to me and worth pursuing. But yes, this is of course work for next year.

We already have that complete flexibility. Any name / ID scheme would just add some conveniences, but not new functionality.

> Both astropy and sherpa have model names as well. It's an established concept.

Well, they don't use names in the way how you suggest we use them, they don't have a SkyModels container class for name-based access. Names are rather used for compound hierarchical models, and I think Astropy and Sherpa don't care about uniqueness of names. Details matter here, that's why we're having this difficult conversation to find our way how to handle names.

> In the __init__ of SkyModels and Datasets we check for the uniqueness of the names

Obviously it's not the way I would have gone, but OK, we can do that. Note that in `Parameters` you added a uniqueness check in `__init__`, and I had to remove it, to allow the use case of creating a model out of two parts that have linked parameters (e.g. two datasets where the same model was on each).

Having a name attribute on each model / dataset, and then a container class models / datasets which is responsible for ensuring uniqueness is intrinsically a bit painful. For the user, because error messages come late (when creating the container), and when implementing you have to either make the container immutable (best by using a Numpy array instead of a Python list), or if you want to keep the list mutable data structure, you have to put the uniqueness change in `__init__`, `append`, `extend`, and any other operation that can change the models. Also you should probably make name a read-only property on model and dataset that can only be set on init. If you look things down this way, then yes, you can achieve this uniqueness.

And I think the most important action item here IMO is: write a short docs section "Dataset, model and parameter names in Gammapy" explaining our approach (like http://xarray.pydata.org/en/stable/faq.html#what-is-your-approach-to-metadata), since that is something that every user and developer has to understand.

Note that for parameters, there's one more change I will implement soon. Instead of
```
class MyModel(Model):
    spam = Parameter("spam", 42)
```
I will change to 
```
class MyModel(Model):
    spam = Parameter(42)
```
and use `__set_name__` (see https://www.python.org/dev/peps/pep-0487/#trait-descriptors) to fill in the Parameter `name` attribute, to guarantee name consistency for parameter objects on models.
It's mostly independent of the discussion on dataset & model names, but it means that there cannot be two parameters with the same name on a model, which IMO is a good constraint to have. (there can still be the same parameter name on submodels, without a problem (see [test](https://github.com/gammapy/gammapy/blob/955a692a6ee78ddf766b06ab4c61a39729521c60/gammapy/modeling/tests/test_model.py#L30))

**Comment by cdeil** [comment]
Discussed again with @adonath @QRemy and @registerrier at the coding sprint.

Plan, to be implemented later today:

- Model and Dataset objects keep a `name` field, we don't have an `id` field instead or in addition.
- `name` is set on `__init__` and is read-only, changing it is not allowed
- by default `name = uuid.uuid4().hex`, which is a random unique 32-char hex string like `'69b6d39f68a048a3aae52753ba999318'`
- the creator can pass a name to `__init__` if they like, or it can come from a file with `read`
- this typically happens for MWL analyses and complex sky models, where users need to keep track of the parts
- on `Datasets` and `SkyModels` init we add a check that all names of the inputs are unique, and raise a `ValueError` if they are not
- in Gammapy we make no attempt to generate names, because it's hard.
- in the Model and Dataset `repr` and `str` we show that `name`, so that users can easily see and debug it. (but they can't change it!).
- on serialisation we rely on unique names, to link datasets and models!!! We don't generate new IDs on write. (doing that is possible, then we wouldn't need to do uniqueness checks in Model and Dataset init, we don't prefer that because it will lead to late fails in user scripts)

Putting `f"{obs_id}"` is not sufficient, e.g. when making LC in 5 min bin, or splitting by event type or ... if we tried to consistently generate nice human-readable names in Gammapy, we would have a big and hard to maintain task on our hand. Global object counters don't work either, because we want to keep names on read and write and counters reset in new Python processes (e.g. write after data reduction, read to model & fit & check output).

In Parameter `name` is not unique, it is e.g.  "amplitude" and there can be many parameter objects with that name in a Gammapy analysis. In the repr we show `hex(id(self))`. We keep this as-is, we don't add a `parameter.id`, we don't need a unique ID for parameters in the current design. We might re-discuss that point tomorrow concerning parameter links in the serialisation. If we want it, the solution is to add a UUID on Parameter `__init__`.

**Comment by adonath** [comment]
Thanks @cdeil for the summary, I have nothing to add. I could go ahead and implement the proposal today or tomorrow.

**Comment by QRemy** [comment]
Implementing the proposal in #2728
